src/util/git_repos.rs

```diff
@@ -162,86 +145,38 @@ impl GitRepoManager {
         }
     }
 
-    /// Runs git rev-parse on the git repo in `repo_dir` to see if `commit_maybe` is a commit in the
-    /// repo.  Returns true if it is an false if it's not.  If it fails for any other reason,
-    /// returns an error.  It should be noted that a true returned from this function does not
-    /// mean `commit_maybe` is not a tag.  This will also return true for tags.  If using this to
-    /// differentiate between tags and commits, `git_show_ref_verify` should be run first to check
-    /// if the string in question is a tag
-    fn git_rev_parse_verify(&self, repo_dir: &Path, commit_maybe: &str) -> Result<bool, Error> {
+    /// Runs git rev-parse on the git repo in `repo_dir` to get the commit hash for `git_ref`.
+    /// Returns the commit hash if `git_ref` is a valid ref, otherwise returns Error::NotFound. If
+    /// it fails for any other reason, returns an error.
+    fn git_rev_parse_verify(&self, repo_dir: &Path, git_ref: &str) -> Result<String, Error> {
         // Run the command
         let output = Command::new("sh")
             .current_dir(repo_dir)
             .arg("-c")
             .arg(format!(
                 "git rev-parse --verify \"{}^{{commit}}\"",
-                commit_maybe
+                git_ref
             ))
             .output()?;
 
-        // If the command was successful, return true
+        // If the command was successful, return the commit hash
         if output.status.success() {
-            Ok(true)
+            Ok(String::from_utf8_lossy(&*output.stdout)
+                .to_string()
+                .trim()
+                .to_string())
         } else {
             let stderr = String::from_utf8_lossy(&*output.stderr).to_string();
             // If stderr matches the message that git spits out if the commit is not valid, return
-            // false
+            // NotFound
             if stderr.trim() == "fatal: Needed a single revision" {
-                return Ok(false);
+                return Err(Error::NotFound(git_ref.to_string()));
```

Comment:
dumb rust question: why do you need to call to string on git_ref which is already a string?

Reply:
git_ref is a &str which is a string slice.  A string slice is basically an immutable reference to an str, a fixed length string in memory which is the rust primitive type for a string.  This is different from a String, which is more like a Java StringBuffer.  You can't create an str; you can only reference its place in memory using an &str.  Because of that, you can never actually own an str, so if you need to own a string, you have to use the String type.  `Error::NotFound` needs to own the value in git_ref, so I'm creating a String from it.

Reply:
Strings in rust are confusing